### DISCUSSION TITLE
Can DND do the react-grid-layout showcase?

### DISCUSSION THREAD

**Discussion by awhitford** [clauderic/dnd-kit#1560] [General]
DND-kit looks preferable over _React Grid Layout_ because its newer and uses TypeScript.
However, their [Showcase](https://react-grid-layout.github.io/react-grid-layout/examples/00-showcase.html) has some desirable features that I'd like to confirm that DND-kit can do the same.
Specifically, can DND-kit support **Resizable** widgets, snap layout, and the **Compaction** layout?

BTW, I did see this [demo](https://master--5fc05e08a4a65d0021ae0bf2.chromatic.com/?path=/story/presets-sortable-grid--variable-sizes) but I don't like how moving a large box into the small box area is basically resizing the box -- instead, the large box move should push the smaller boxes over.

**Comment by fanselau** [comment]
Hey, have you ever found a solution or made progress on the problem. Would also be interested into what would be the way to implement a custom layouting in dnd-kit (akin to gridstack.js). And how we could implement repositioning during drag.

**Comment by eleung** [comment]
I built [snapgrid](https://snapgrid.dev) for pretty much this exact gap, a react-grid-layout alternative on top of dnd-kit (@dnd-kit/react). It's basically RGL but with dnd-kit's DNA (uses useDroppable and useSortable internally). Supports a few extra things RGL does not as well (cross-grid, keyboard support, headless-first hook based api).

It covers what's asked here: resizable widgets with min/max, snap-to-grid, and pluggable compaction (vertical/horizontal/masonry/gravity/shelf). @awhitford, dragging a big tile pushes the others out of the way rather than resizing into the gap (just like RGL). @fanselau, the engine is headless and framework-agnostic (just like the new dnd-kit rewrite), with a live placeholder during drag, so custom layouts are the main use case.

Disclosure: it's new and I'm the author, so not battle-tested like RGL yet, but it's a project I'm taking seriously, it's MIT and there are live demos to try:

[GitHub Repo](https://github.com/eleung/snapgrid)
[Live Examples](https://snapgrid.dev/react/examples/)
[Larger Showcases](https://snapgrid.dev/showcase/)

**Comment by davideagosti-dev** [comment]
@eleung Interesting approach with SnapGrid, particularly keeping the engine headless and making compaction pluggable.

I'm curious how you're thinking about constraints beyond geometric min/max.

Have you considered allowing a tile to express a “useful minimum” separately from its hard minimum? For example, a chart might technically resize down to 2 columns, but below 4 columns its labels/data become practically unusable.

It raises an interesting layout question: should the engine consider that state valid, invalid, or valid-but-degraded when deciding how to rearrange surrounding tiles?
